keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2144 @@
+the `drainTargets`. For example, if we observe the following NodeMaintenance:
+
+```yaml
+apiVersion: v1alpha1
+kind: NodeMaintenance
+...
+status:
+  nodeStatuses:
+    - nodeRef:
+        name: six
+      drainTargets:
+        - podPriority: 2147483647
+          podType: Default
+        - podPriority: 5000
+          podType: DaemonSet
+  ...
+```
+
+To register the daemonset controller as a controller evacuator, it should set the following
+annotation on its own pods.
+
+```yaml
+evacuation.coordination.k8s.io/daemonset.apps.k8s.io: "10000/controller"
+```
+
+The controller should respond to the Evacuation object when it observes its own class
+(`daemonset.apps.k8s.io`) in `.status.activeEvacuatorClass`.
+
+For the above node maintenance, the controller should not react to Evacuations of DaemonSet pods
+with a priority greater than 5000. This state should not normally occur, as Evacuation requests
+should be coordinated with NodeMaintenance. If it does occur, we should not encourage this flow by
+updating the `.status.ActiveEvacuatorCompleted` field, although it is required to update this field
+for normal workloads.
+
+If the DaemonSet pods have a priority equal to or less than 5000, the Evacuation status should be
+updated appropriately as follows, and the targeted pod should be deleted by the daemonset
+controller:
+
+```yaml
+apiVersion: v1alpha1
+kind: Evacuation
+metadata:
+  finalizers:
+    evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io
+  labels:
+    app: critical-ds
+  name: ae9b4bc6-e4ca-4f8e-962b-2d4459b1f684-critical-ds-5nxjs
+  namespace: critical-workloads
+spec:
+  podRef:
+    name: critical-ds-5nxjs
+    uid:  ae9b4bc6-e4ca-4f8e-962b-2d4459b1f684
+  progressDeadlineSeconds: 1800
+  evacuators:
+    - evacuatorClass: daemonset.apps.k8s.io
+      priority: 10000
+      role: controller
+status:
+  activeEvacuatorClass: daemonset.apps.k8s.io
+  activeEvacuatorCompleted: false
+  evacuationProgressTimestamp: "2024-04-22T21:40:32Z"
+  expectedEvacuationFinishTime: "2024-04-22T21:41:32Z" # now + terminationGracePeriodSeconds:
+  failedEvictionCounter: 0
+  message: "critical-ds is terminating the pod due to node maintenance (OS upgrade)."
+  conditions: []
+```
+
+Once the pod is terminated and removed from the node, it should not be re-scheduled on the node by
+the deamonset controller until the node maintenance is complete.
+
+### Kubelet: Graceful Node Shutdown
+
+The current Graceful Node Shutdown feature has a couple of drawbacks when compared to
+NodeMaintenance:
+- It is application agnostic as it only provides a static grace period before the shutdown based on
+  priority. This does not always give the application enough time to react and can lead to data
+  loss or application availability loss.
+- The DaemonSet pods may be running important services (critical priority) that should be available
+  even during part of the shutdown. The daemon set controller does not have the observability of
```

Comment:
A recently filled issue also highlighted a problem between the Taint manager and Graceful Node Shutdown, which, similarly to the problem with the DaemonSet controller, can potentially interfere with the graceful termination. See:

- https://github.com/kubernetes/kubernetes/issues/124448
- https://github.com/kubernetes/kubernetes/issues/72129

Lack of information or internet (if you wish) exchange between Graceful Node Shutdown manager and rest of the platform is definitely something that seems to be a problem more often than not.

Reply:
+1, I haved added a mention of these under `#8`  in [Motivation](https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#motivation). Also mentioned in the [non-goals](https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#non-goals).

I am not really sure if we can stop pod termination mechanisms (e.g. taint eviction) for pods that are under NodeMaintenance/Evacuation. Seems like a backwards incompatible change. @kwilczynski @sftim thoughts?